File: love2d/src/engine/state_machine.lua
local State = {}
State.__index = State

function State.new(state_machine, run_manager)
    local self = setmetatable({}, State)
    self.state_machine = state_machine
    self.run_manager = run_manager
    return self
end

function State:enter(kwargs) end
function State:exit() end
function State:update(dt) end
function State:draw() end
function State:keypressed(key, scancode, isrepeat) end
function State:textinput(text) end
function State:mousepressed(x, y, button, istouch, presses) end

local StateMachine = {}
StateMachine.__index = StateMachine

function StateMachine.new()
    local self = setmetatable({}, StateMachine)
    self.states = {}
    self.current_state = nil
    return self
end

function StateMachine:add_state(name, state)
    self.states[name] = state
end

function StateMachine:change_state(name, kwargs)
    if self.current_state then
        self.current_state:exit()
    end
    self.current_state = self.states[name]
    if self.current_state then
        self.current_state:enter(kwargs or {})
    end
end

function StateMachine:update(dt)
    if self.current_state then
        self.current_state:update(dt)
    end
end

function StateMachine:draw()
    if self.current_state then
        self.current_state:draw()
    end
end

function StateMachine:keypressed(key, scancode, isrepeat)
    if self.current_state then
        self.current_state:keypressed(key, scancode, isrepeat)
    end
end

function StateMachine:textinput(text)
    if self.current_state then
        self.current_state:textinput(text)
    end
end

function StateMachine:mousepressed(x, y, button, istouch, presses)
    if self.current_state then
        self.current_state:mousepressed(x, y, button, istouch, presses)
    end
end

return {
    State = State,
    StateMachine = StateMachine
}

File: src/engine/state_machine.py
import pygame

class State:
    def __init__(self, state_machine, run_manager):
        self.state_machine = state_machine
        self.run_manager = run_manager

    def enter(self, **kwargs):
        """Called when transitioning into this state."""
        pass

    def exit(self):
        """Called when transitioning out of this state."""
        pass

    def handle_events(self, events):
        """Handles pygame events."""
        pass

    def update(self, dt):
        """Updates state logic. dt is in seconds."""
        pass

    def draw(self, surface):
        """Renders the state graphics to the surface."""
        pass


class StateMachine:
    def __init__(self):
        self.states = {}
        self.current_state = None

    def add_state(self, name, state):
        self.states[name] = state

    def change_state(self, name, **kwargs):
        """Cleanly transition from current state to the specified state."""
        if self.current_state:
            self.current_state.exit()
        
        self.current_state = self.states[name]
        self.current_state.enter(**kwargs)

    def handle_events(self, events):
        if self.current_state:
            self.current_state.handle_events(events)

    def update(self, dt):
        if self.current_state:
            self.current_state.update(dt)

    def draw(self, surface):
        if self.current_state:
            self.current_state.draw(surface)
